### DISCUSSION TITLE
Some question about Line features

### DISCUSSION THREAD

**Discussion by LeeLiang123** [lagadic/visp#1182] [Q&A]
Hello Professor, thank you very much for providing such a useful open source code,but it's difficult for me to figure out the C++ code.So, I am having some question about visual servo using line features.
First is how to calculate the equation of a plane to which the line belongs?I tried to solve the equation using the information on the line detected by the Hough transform and converting the pixel coordinates to points in the camera frame to solve the equation, but it didn't seem to work.
Another question is which parameters are used to calculate the error? For example, four straight lines are detected, so the dimension of the error should be 4 times 6, is that correct?
Your reply will be of great help to me, thanks！

**Comment by LeeLiang123** [comment]
@fspindle

**Reply by fspindle** [reply]
From the image processing you have to compute the parameters of the line in the image (rho, theta) which could be used as the visual features in the visual servoing scheme, meaning that you don't have to calculate the equation of the planes.

For one line since the parameters are (rho, theta), the error is a vector of dimension 2 (rho - rho^*, theta-theta^*) where (rho^*, theta^*) are the parameters of the desired line.

Having a look at [vpFeatureLine](https://visp-doc.inria.fr/doxygen/visp-daily/classvpFeatureLine.html) documentation you will find a lot of examples.

**Reply by LeeLiang123** [reply]
Hello again! There is one more question i haven't understood yet.The A, B, C, D parameters are needed to compute the interaction matrix associated to a visual feature.For a line we just need a plane equation to get A,B,C,D. But how should we go about solving this equation. I try to solve this equation using three points, but I don't know which coordinate system to use. What's more, assuming there are four lines, do we need to solve A1, B1, C1, D1...A4, B4, C4, D4. @fspindle

**Reply by fspindle** [reply]
You don't have to estimate the equation of the plane. As in this [example](https://visp-doc.inria.fr/doxygen/visp-daily/servoAfma6Line2DCamVelocity_8cpp-example.html#_a28) line 144, you can specify the equation of the two planes in the object frame (^oX=0, ^oZ=0) at the desired position. Then specify the ^cM_o matrix corresponding to the desired pose of the camera wrt the object (line 145), and project the line (line 146) to update the parameters of the line in the camera frame at the desired position you want to reach.

**Reply by LeeLiang123** [reply]
Thank you Professor,we did it!But one thing I haven't figured out is how the parameters of the plane equation affect the robot's motion.
The translation and rotation of the robot in the x and y directions can be controlled by the parameters rho and theta,then the parameters of the plane equation affect the rotation and translation of the robot in the z direction,is that correct? @fspindle
